File: ein.py/src/ein_bot/inference/contradiction.py
"""Contradiction detector — S1.4.1 / P1.4 (+ S1.5.4a Part 2 direct ⊥).

Scans the KB for two contradiction shapes, both emitting
:class:`Contradiction` records the hypothesis loop (P1.5) and trace
renderer (P1.6) consume:

1. **pair** — a same-layer ``(X, (not X))`` pair. The original P1.4
   shape: a rule asserted ``(not X)`` and ``X`` is also in the KB
   (typically seeded by a hypothesis fork). ``positive`` and
   ``negative`` are both populated.
2. **direct** — a ``(false)`` fact (S1.5.4a Part 2): a rule
   asserted contradiction directly without going through the
   self-negation idiom. ``positive`` is ``None``; ``negative`` is
   the ``(false …)`` fact itself, which carries the firing's
   provenance for unsat-core walks.

Both encode a **branch failure** under M1's append-only KB model:
there's no way to retract a fact once asserted, so the only
resolution is for whatever caused the conflict (typically a
hypothesis fork) to be unwound. P1.4 ships the *detector*; P1.5
ships the *unwinder*.

The detector is a pure scan over the existing fact set — no new
entity kinds, no indexes, no incremental machinery. The
Saturator's append-only ``_index_fact`` already keeps
``_facts_by_relation`` current; this module just reads it.

Same-layer vs cross-layer (applies to the ``pair`` shape only):

- A REASONING-layer ``(not X)`` derived from ``type-exclusivity``
  is the *expected* output of saturation. If the matching
  REASONING-layer positive ``X`` is also present (e.g. asserted
  speculatively by a hypothesis), the pair is a branch failure.
- A cross-layer pair (FACT-layer X + REASONING-layer ``(not X)``)
  is **not** a contradiction in this design — layer separation is
  the engine's way of marking "different epistemic statuses".
  Both layers being co-present is by construction (the matcher
  consults them jointly), and a FACT-layer positive shouldn't
  cause a derived negative to flag it as broken. P1.5 will
  introduce branch-scoped layers; the cross-source case becomes
  meaningful there.

Direct ⊥ dedup caveat: ``Fact`` identity is by ``(relation, args)``,
so the second-and-later ``(false)`` firings within a single
fork collapse onto the first — only the first firing's provenance
is preserved. That's sufficient for "is this branch dead?" and
for back-prop to identify the responsible hypothesis (the first
firing's premise chain reaches it). Promote to ``(false <witness>)``
with per-firing args if a future puzzle needs all parallel
contradictions reported.
"""
from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass
from typing import Literal

from ein_bot.kb.entities import Fact, Layer
from ein_bot.kb.store import KnowledgeBase

# ── Contradiction record ──────────────────────────────────────────


@dataclass(frozen=True)
class Contradiction:
    """A contradiction found by the detector — pair or direct.

    For ``kind="pair"``: ``positive`` is the ``X`` fact and
    ``negative`` is the wrapping ``(not X)`` fact;
    ``negative.args[0]`` is the inner positive ``Fact``.

    For ``kind="direct"``: ``positive`` is ``None`` and
    ``negative`` is the ``(false …)`` fact whose provenance walks
    back to the firing's premises.

    Use :attr:`witness` to get whichever fact is appropriate as
    the seed for an unsat-core walk — ``positive`` if present,
    otherwise ``negative``.

    The dataclass is frozen so it can live in sets if a caller
    wants to dedupe a multi-pass scan.
    """
    positive: Fact | None  # the X fact, or None for direct ⊥
    negative: Fact         # (not X) wrapper OR the (false …) fact itself
    layer: Layer           # the layer the contradiction lives in
    kind: Literal["pair", "direct"] = "pair"

    @property
    def witness(self) -> Fact:
        """The fact whose derivation DAG seeds the unsat-core walk.

        For pair contradictions: the positive ``X`` (its DAG walks
        back to the premises that derived it; the negative is
        typically a structural rule firing whose DAG is shallow).
        For direct contradictions: the ``(false …)`` fact, whose
        DAG IS the firing chain of the rule that emitted ⊥.
        """
        return self.positive if self.positive is not None else self.negative


# ── Detector ──────────────────────────────────────────────────────


class ContradictionDetector:
    """Scanner over ``kb.facts`` looking for ``(X, (not X))`` pairs.

    Construct once per KB; call :meth:`detect` whenever a pair
    snapshot is needed (typically: after a saturation cycle, before
    a hypothesis-loop branch decision). The detector keeps no
    state between calls — repeated scans are independent.

    Complexity: O(|`not`-headed facts|) per scan. On Zebra
    (~120 negatives after saturation), each iteration does one
    O(deg(rel)) lookup in :attr:`KnowledgeBase._facts_by_relation`
    — bounded by the number of facts with that relation name.
    Sub-millisecond on Zebra-scale.
    """

    def __init__(self, kb: KnowledgeBase) -> None:
        self.kb = kb

    # ── Public API ────────────────────────────────────────────────

    def detect(self) -> tuple[Contradiction, ...]:
        """All ``(X, (not X))`` same-layer pairs across the KB."""
        return tuple(self._iter())

    def detect_layer(self, layer: Layer) -> tuple[Contradiction, ...]:
        """Pairs scoped to a single layer."""
        return tuple(c for c in self._iter() if c.layer is layer)

    def has_contradiction(self) -> bool:
        """Short-circuit yes/no — stops on the first pair found.

        Faster than ``bool(detect())`` when the KB has many
        negative facts and only the existence-question matters
        (e.g. P1.5 deciding whether to retract a branch).
        """
        for _ in self._iter():
            return True
        return False

    # ── Iterator (single source of truth) ─────────────────────────

    def _iter(self) -> Iterator[Contradiction]:
        """Yield each contradiction once. Used by all three public
        methods to keep the algorithm in one place.

        Two shapes:
        - **pair** — walks ``_facts_by_relation['not']`` and matches
          each ``(not X)`` against an existing positive ``X``.
        - **direct** — walks ``_facts_by_relation['false']`` and
          yields one ``Contradiction(kind='direct', positive=None)``
          per ``(false …)`` fact.
        """
        # Direct ⊥ — ship S1.5.4a Part 2.
        for false_fact in self.kb._facts_by_relation.get("false", ()):
            yield Contradiction(
                positive=None,
                negative=false_fact,
                layer=false_fact.layer,
                kind="direct",
            )

        # Pair — the original P1.4 shape.
        not_facts = self.kb._facts_by_relation.get("not", ())
        for negative in not_facts:
            if not negative.args:
                continue  # malformed; skip defensively
            inner = negative.args[0]
            if not isinstance(inner, Fact):
                # Q40 / R9 widening: the inner of a `(not …)` is
                # expected to be a Fact. If it's a string/int,
                # there's no positive proposition to match against
                # — the engine's loader / matcher should never
                # produce this shape, but we tolerate it.
                continue
            positive = self.kb._fact_by_id(inner.relation_name, inner.args)
            if positive is None:
                continue
            if positive.layer is not negative.layer:
                # Cross-layer pair — by design, NOT a contradiction.
                # See module docstring.
                continue
            yield Contradiction(
                positive=positive,
                negative=negative,
                layer=positive.layer,
                kind="pair",
            )


__all__ = ["Contradiction", "ContradictionDetector"]

File: ein.rs/crates/ein-infer/src/contradiction.rs
//! Contradiction detection — the two shapes, and the incremental check.
//!
//! 1. **pair** — an `(X, (not X))` pair. A rule asserted `(not X)` and `X` is
//!    also in the KB.
//! 2. **direct** — a `(false)` fact: a rule asserted contradiction outright
//!    rather than through the self-negation idiom.
//!
//! Both encode a **branch failure** under M1's append-only KB: there is no way
//! to retract a fact, so the only resolution is for whatever caused the
//! conflict — typically a hypothesis fork — to be unwound.
//!
//! **How a fact got into the KB is irrelevant** (S1.22.1b). A KB holding both
//! `X` and `(not X)` is inconsistent whether `X` is a stated clue, a
//! background assumption or a derivation, so every such pair is reported.
//! Until S1.22.1b the detector skipped pairs whose facts sat in different
//! knowledge layers, and that was a soundness bug: a flatly inconsistent
//! puzzle was accepted in silence.
//!
//! The port's change is the *lookup*, not the algorithm
//! ([design/06](../../../../docs/history/m1a_rust/design/06_saturation.md) §6):
//! ein.py's `(rn, args) in kb._negated_facts` and `_fact_by_id(X)` become two
//! bit tests. The result **order** is preserved — direct ⊥ first, then pairs
//! in extent order — because it reaches the unsat core and the trace.

use ein_core::{FactId, Kb, Terms};

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Kind {
    Pair,
    Direct,
}

impl Kind {
    pub fn as_str(self) -> &'static str {
        match self {
            Kind::Pair => "pair",
            Kind::Direct => "direct",
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub struct Contradiction {
    /// The `X` fact, or `None` for a direct ⊥.
    pub positive: Option<FactId>,
    /// The `(not X)` wrapper, or the `(false …)` fact itself.
    pub negative: FactId,
    pub kind: Kind,
}

impl Contradiction {
    /// The fact whose derivation DAG seeds the unsat-core walk.
    ///
    /// For a pair, the positive `X`: its DAG walks back to the premises that
    /// derived it, where the negative is typically a structural rule firing
    /// with a shallow one. For a direct ⊥, the `(false …)` fact, whose DAG
    /// *is* the firing chain of the rule that emitted it.
    pub fn witness(&self) -> FactId {
        self.positive.unwrap_or(self.negative)
    }
}

/// Every contradiction in the KB, direct ⊥ first.
pub fn detect(kb: &Kb, terms: &Terms) -> Vec<Contradiction> {
    let mut out = Vec::new();
    for false_fact in kb.facts_of(terms.kernel.r#false) {
        out.push(Contradiction {
            positive: None,
            negative: false_fact,
            kind: Kind::Direct,
        });
    }
    for negative in kb.facts_of(terms.kernel.not) {
        // Q40 / R9's widening: the inner of a `(not …)` is expected to be a
        // fact. A string or an int leaves no positive proposition to match
        // against — the loader and matcher never produce that shape, but it is
        // tolerated rather than asserted.
        let Some(inner) = terms.facts.args(negative).first().and_then(|v| v.as_fact()) else {
            continue;
        };
        if kb.contains(inner) {
            out.push(Contradiction {
                positive: Some(inner),
                negative,
                kind: Kind::Pair,
            });
        }
    }
    out
}

/// Short-circuit yes/no — stops on the first contradiction found.
pub fn has_contradiction(kb: &Kb, terms: &Terms) -> bool {
    if kb.n_facts_of(terms.kernel.r#false) > 0 {
        return true;
    }
    kb.facts_of(terms.kernel.not).any(|negative| {
        terms
            .facts
            .args(negative)
            .first()
            .and_then(|v| v.as_fact())
            .is_some_and(|inner| kb.contains(inner))
    })
}

/// True iff `fact`, just written to `kb`, makes it inconsistent.
///
/// The **incremental** dual of [`detect`]: the scan asks "does this KB hold a
/// contradiction?", this asks "did *this fact* create one?". On a KB that was
/// consistent before `fact` landed the two agree exactly, and that equivalence
/// is what lets fork saturation fail fast (S1.9.E23): the KB is append-only,
/// so a contradiction can only be *created* by an insertion and can never be
/// retracted, so checking each derived fact as it lands finds every death a
/// post-fixpoint scan would — about 2.5 k firings earlier on a dying zebra2
/// fork, which is ~88 % of its saturation.
pub fn contradicts(kb: &Kb, terms: &Terms, fact: FactId) -> bool {
    let rel = terms.facts.rel(fact);
    if rel == terms.kernel.r#false {
        return true; // direct ⊥
    }
    if rel == terms.kernel.not {
        // `fact` is `(not X)` — dead iff the positive `X` is believed.
        return terms
            .facts
            .args(fact)
            .first()
            .and_then(|v| v.as_fact())
            .is_some_and(|inner| kb.contains(inner));
    }
    // `fact` is positive — dead iff some `(not fact)` is believed.
    kb.is_negated(fact)
}

#[cfg(test)]
mod tests {
    use super::*;
    use ein_ir::{Ast, from_ir::load, parse};

    fn kb_of(src: &str) -> (Terms, Kb) {
        let mut ast = Ast::new();
        let mut terms = Terms::new();
        let forms = parse(&mut ast, src, Some("<test>")).expect("parses");
        let kb = load(&mut ast, &mut terms, &forms, None).expect("loads");
        (terms, kb)
    }

    #[test]
    fn a_pair_and_a_direct_are_both_found_direct_first() {
        let (terms, kb) = kb_of(
            "(relation p Thing)\n(p one)\n(not (p one))\n(false)\n(p two)\n(not (p three))\n",
        );
        let found = detect(&kb, &terms);
        assert_eq!(found.len(), 2, "{found:?}");
        assert_eq!(found[0].kind, Kind::Direct);
        assert_eq!(found[0].positive, None);
        assert_eq!(found[1].kind, Kind::Pair);
        // `(not (p three))` has no positive to pair with, and `(p two)` has no
        // negation — neither is a contradiction on its own.
        assert!(has_contradiction(&kb, &terms));
    }

    #[test]
    fn the_incremental_check_agrees_with_the_scan() {
        let (terms, kb) = kb_of("(relation p Thing)\n(p one)\n(not (p one))\n(p two)\n");
        // Every fact the KB holds, checked both ways. The scan is over the
        // whole KB, so on a consistent-until-now KB the two agree fact by fact.
        let flagged: Vec<bool> = kb.facts().map(|f| contradicts(&kb, &terms, f)).collect();
        assert_eq!(
            flagged.iter().filter(|x| **x).count(),
            2,
            "`(p one)` and `(not (p one))` each create the pair"
        );
        assert!(
            detect(&kb, &terms).len() == 1,
            "the pair is one contradiction"
        );
    }

    #[test]
    fn a_consistent_kb_reports_nothing() {
        let (terms, kb) = kb_of("(relation p Thing)\n(p one)\n(not (p two))\n");
        assert!(detect(&kb, &terms).is_empty());
        assert!(!has_contradiction(&kb, &terms));
        for f in kb.facts() {
            assert!(!contradicts(&kb, &terms, f));
        }
    }
}
